xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderClientSslContextProvider.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.certprovider;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.envoyproxy.envoy.config.core.v3.Node;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CertificateValidationContext;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CommonTlsContext;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CommonTlsContext.CombinedCertificateValidationContext;
+import io.grpc.netty.GrpcSslContexts;
+import io.grpc.xds.Bootstrapper.CertificateProviderInfo;
+import io.grpc.xds.EnvoyServerProtoData.UpstreamTlsContext;
+import io.grpc.xds.internal.sds.trust.SdsTrustManagerFactory;
+import io.netty.handler.ssl.SslContextBuilder;
+import java.security.cert.CertStoreException;
+import java.security.cert.X509Certificate;
+import java.util.Map;
+
+/** A client SslContext provider using CertificateProviderInstance to fetch secrets. */
+final class CertProviderClientSslContextProvider extends CertProviderSslContextProvider {
+
+  private CertProviderClientSslContextProvider(
+      Node node,
+      Map<String, CertificateProviderInfo> certProviders,
+      CommonTlsContext.CertificateProviderInstance certInstance,
+      CommonTlsContext.CertificateProviderInstance rootCertInstance,
+      CertificateValidationContext staticCertValidationContext,
+      UpstreamTlsContext upstreamTlsContext,
+      CertificateProviderStore certificateProviderStore) {
+    super(
+        node,
+        certProviders,
+        certInstance,
+        checkNotNull(rootCertInstance, "Client SSL requires rootCertInstance"),
+        staticCertValidationContext,
+        upstreamTlsContext,
+        certificateProviderStore);
+  }
+
+  @Override
+  public SslContextBuilder getSslContextBuilder(
+          CertificateValidationContext certificateValidationContextdationContext)
+      throws CertStoreException {
+    SslContextBuilder sslContextBuilder =
+        GrpcSslContexts.forClient()
+            .trustManager(
+                new SdsTrustManagerFactory(
+                    savedTrustedRoots.toArray(new X509Certificate[0]),
+                    certificateValidationContextdationContext));
+    if (isMtls()) {
+      sslContextBuilder.keyManager(savedKey, savedCertChain);
```

Comment:
`savedKey` and `savedCertChain` are very likely `null` all the time, because even they are set by `updateCertificate()`,  `updateCertificate()` may  immediately and automatically trigger `clearKeysAndCerts()`.

Reply:
Note that the [`saved*`](https://github.com/grpc/grpc-java/blob/2cf81bc5072fde5062d1738c2a2f336f602b389f/xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderSslContextProvider.java#L39) values are used to generate the [`sslContext`](https://github.com/grpc/grpc-java/blob/2cf81bc5072fde5062d1738c2a2f336f602b389f/xds/src/main/java/io/grpc/xds/internal/sds/DynamicSslContextProvider.java#L41) and after that all the `saved*` fields are set to `null`. So you are right most of the time the `saved*` values are null but `sslContext` will be non-null (if it has been generated once) which is what is needed by all the consumers of this class.

The reason for setting the `saved*` values to null is because of the separate calls `updateCertificate` and `updateTrustedRoots` from the CertificateProvider. I can build an SslContext for a given state (mTLS, clientTls or serverTls) only when all the values are in. Without assuming any order for `updateCertificate` and `updateTrustedRoots` calls the best (and only?) way to do this is to check for the required `saved*` values to be non-null and setting them to null (after building the `sslContext`) for the next cycle of `updateCertificate` and `updateTrustedRoots`.